### PR TITLE
[test] Introduce workaround for static linking tests on Dom

### PR DESCRIPTION
**Pull request by teojgo**
* Workaround for Cray Bug #255707.

### CODE REVIEW

File: cscs-checks/libraries/io/hdf5_compile_run.py [hunk 1]
```diff
@@ -74,3 +75,13 @@ def __init__(self, lang, linkage):
...
+
+    @rfm.run_after('setup')
+    def set_linker_variables(self):
+        # FIXME: static compilation yields a link error in case of
+        # PrgEnv-cray(Cray Bug #255707)
+        if (self.linkage == 'static' and
+            self.current_system.name == 'dom' and
+            self.current_environ.name == 'PrgEnv-cray'):
+            self.variables = {'LINKER_X86_64': '/usr/bin/ld',
+                              'LINKER_AARCH64': '=/usr/bin/ld'}
```
Comment (vkarak) [inline cscs-checks/libraries/io/hdf5_compile_run.py]: What's the `=` about?

File: cscs-checks/libraries/io/hdf5_compile_run.py [hunk 2]
```diff
@@ -74,3 +75,13 @@ def __init__(self, lang, linkage):
 
         self.maintainers = ['SO', 'RS']
         self.tags = {'production', 'craype'}
+
+    @rfm.run_after('setup')
```
Comment (vkarak) [inline cscs-checks/libraries/io/hdf5_compile_run.py]:
```suggestion
    @rfm.run_before('compile')
```

File: cscs-checks/libraries/io/hdf5_compile_run.py [hunk 3]
```diff
@@ -74,3 +75,13 @@ def __init__(self, lang, linkage):
...
+    def set_linker_variables(self):
```
Comment (vkarak) [inline cscs-checks/libraries/io/hdf5_compile_run.py]:
```suggestion
    def cray_linker_workaround(self):
```

File: cscs-checks/libraries/io/hdf5_compile_run.py [hunk 4]
```diff
@@ -74,3 +75,13 @@ def __init__(self, lang, linkage):
...
+        # FIXME: static compilation yields a link error in case of
+        # PrgEnv-cray(Cray Bug #255707)
+        if (self.linkage == 'static' and
+            self.current_system.name == 'dom' and
+            self.current_environ.name == 'PrgEnv-cray'):
```
Comment (vkarak) [inline cscs-checks/libraries/io/hdf5_compile_run.py]: Inverse the check here and simply return. Then set the variables unindented.

File: cscs-checks/libraries/io/netcdf_compile_run.py [hunk 5]
```diff
@@ -43,9 +43,16 @@ def __init__(self, lang, linkage):
...
     @rfm.run_before('compile')
     def setflags(self):
-        environ = self.current_environ
+        # FIXME: static compilation yields a link error in case of
+        # PrgEnv-cray(Cray Bug #255707)
+        if (self.linkage == 'static' and
+            self.current_system.name == 'dom' and
+            self.current_environ.name == 'PrgEnv-cray'):
+            self.variables = {'LINKER_X86_64': '/usr/bin/ld',
+                              'LINKER_AARCH64': '=/usr/bin/ld'}
```
Comment (vkarak) [inline cscs-checks/libraries/io/netcdf_compile_run.py]: Put everything in a separate decorated function. ReFrame will execute all the hooks. Use the same name as the one proposed above.

File: cscs-checks/libraries/math/scalapack_compile_run.py [hunk 6]
```diff
@@ -27,13 +28,26 @@ def __init__(self, linkage):
         self.maintainers = ['CB', 'LM']
         self.tags = {'production', 'external-resources'}
 
+    @rfm.run_before('compile')
+    def set_linker_variables(self):
```
Comment (vkarak) [inline cscs-checks/libraries/math/scalapack_compile_run.py]: Same comments as above.

File: cscs-checks/libraries/math/scalapack_compile_run.py [hunk 7]
```diff
@@ -27,13 +28,26 @@ def __init__(self, linkage):
...
 
 @rfm.required_version('>=2.14')
 @rfm.parameterized_test(['static'], ['dynamic'])
 class ScaLAPACKSanity(ScaLAPACKTest):
     def __init__(self, linkage):
         super().__init__(linkage)
         self.sourcepath = 'scalapack_compile_run.f'
+        if linkage == 'static':
+            self.variables['LINKER_X86_64'] = '/usr/bin/ld'
+            self.variables['LINKER_AARCH64'] = '/usr/bin/ld'
```
Comment (vkarak) [inline cscs-checks/libraries/math/scalapack_compile_run.py]:
```python
self.variables.update{{'LINKER_X86_64': '/usr/bin/ld', 'LINKER_AARCH64': '/usr/bin/ld'})
```

File: cscs-checks/prgenv/helloworld.py [hunk 8]
```diff
@@ -81,8 +79,19 @@ def num_ranks(match):
...
+
     @rfm.run_before('compile')
     def setflags(self):
+        # FIXME: static compilation yields a link error in case of
+        # PrgEnv-cray(Cray Bug #255707)
+        if (self.linkage == 'static' and
+            self.current_system.name == 'dom' and
+            self.current_environ.name.startswith('PrgEnv-cray')):
+            self.variables = {'LINKER_X86_64': '/usr/bin/ld',
+                              'LINKER_AARCH64': '=/usr/bin/ld'}
```
Comment (vkarak) [inline cscs-checks/prgenv/helloworld.py]: Again, all this in a separate hooked function.

File: cscs-checks/prgenv/helloworld.py [hunk 9]
```diff
@@ -155,9 +164,7 @@ def __init__(self, lang, linkage):
 
         # On SLURM there is no need to set OMP_NUM_THREADS if one defines
         # num_cpus_per_task, but adding for completeness and portability
-        self.variables = {
-            'OMP_NUM_THREADS': str(self.num_cpus_per_task)
-        }
+        self.variables['OMP_NUM_THREADS']: str(self.num_cpus_per_task)
```
Comment (vkarak) [inline cscs-checks/prgenv/helloworld.py]: Keep this as it is and define the workaround as a separate hook after this one. The workaround should do `self.variables.update(...)`.

File: cscs-checks/prgenv/helloworld.py [hunk 10]
```diff
@@ -210,6 +217,4 @@ def __init__(self, lang, linkage):
 
         # On SLURM there is no need to set OMP_NUM_THREADS if one defines
         # num_cpus_per_task, but adding for completeness and portability
-        self.variables = {
-            'OMP_NUM_THREADS': str(self.num_cpus_per_task)
-        }
+        self.variables['OMP_NUM_THREADS'] = str(self.num_cpus_per_task)
```
Comment (vkarak) [inline cscs-checks/prgenv/helloworld.py]: Revert this change.

File: cscs-checks/prgenv/helloworld.py [hunk 11]
```diff
@@ -156,7 +169,7 @@ def __init__(self, lang, linkage):
         # On SLURM there is no need to set OMP_NUM_THREADS if one defines
         # num_cpus_per_task, but adding for completeness and portability
         self.variables = {
-            'OMP_NUM_THREADS': str(self.num_cpus_per_task)
+            'OMP_NUM_THREADS':  str(self.num_cpus_per_task)
```
Comment (vkarak) [inline cscs-checks/prgenv/helloworld.py]:
```suggestion
            'OMP_NUM_THREADS': str(self.num_cpus_per_task)
```

File: cscs-checks/libraries/io/hdf5_compile_run.py [hunk 12]
```diff
@@ -74,3 +75,15 @@ def __init__(self, lang, linkage):
...
+    @rfm.run_before('compile')
+    def cray_linker_workaround(self):
+        # FIXME: static compilation yields a link error in case of
+        # PrgEnv-cray(Cray Bug #255707)
+        if not (self.linkage == 'static' and
+                self.current_system.name == 'dom' and
+                self.current_environ.name == 'PrgEnv-cray'):
+            return
+
+        self.variables = {'LINKER_X86_64': '/usr/bin/ld',
```
Comment (victorusu) [inline cscs-checks/libraries/io/hdf5_compile_run.py]: @teojgo, I think it is easier on the Cray to define
`{'ALT_LINKER': '/usr/bin/ld'}`. 
This way you do not have to set the architecture specific ones.

Comment (vkarak) [reply cscs-checks/libraries/io/hdf5_compile_run.py]: @teojgo If that works, I agree with @victorusu

Comment (teojgo) [reply cscs-checks/libraries/io/hdf5_compile_run.py]: @victorusu, @vkarak it seems to work fine, thanks.

**Comment by vkarak** [comment]
@teojgo It seems that this PR breaks the tests on Kesch. Can you  check?

**Comment by vkarak** [comment]
@jenkins-cscs retry none
